test/functional/api/keymap_spec.lua

```diff
@@ -0,0 +1,243 @@
+
+local helpers = require('test.functional.helpers')(after_each)
+local clear = helpers.clear
+local eq = helpers.eq
+local eval = helpers.eval
+local funcs = helpers.funcs
+local nvim = helpers.nvim
+local source = helpers.source
+
+local function local_copy(t)
+  local copy = {}
+  for k,v in pairs(t) do
+    copy[k] = v
+  end
+  return copy
+end
+
+describe('get_keymap', function()
+  before_each(clear)
+
+  -- Basic mapping and table to be used to describe results
+  local foo_bar_string = 'nnoremap foo bar'
+  local foo_bar_map_table = {
+      lhs='foo',
+      silent=0,
+      rhs='bar',
+      expr=0,
+      sid=0,
+      buffer=0,
+      nowait=0,
+      mode='n',
+      noremap=1,
+    }
+
+  it('returns empty list when no map', function()
+    eq({}, nvim('get_keymap', 'n'))
+  end)
+
+  it('returns list of all applicable mappings', function()
+    nvim('command', foo_bar_string)
+    -- Only one mapping available
+    -- Should be the same as the dictionary we supplied earlier
+    -- and the dictionary you would get from maparg
+    -- since this is a global map, and not script local
+    eq({foo_bar_map_table}, nvim('get_keymap', 'n'))
+    eq(
+      {funcs.maparg('foo', 'n', false, true)},
+      nvim('get_keymap', 'n')
+    )
+
+    -- Add another mapping
+    nvim('command', 'nnoremap foo_longer bar_longer')
+    local foolong_bar_map_table = local_copy(foo_bar_map_table)
+    foolong_bar_map_table['lhs'] = 'foo_longer'
+    foolong_bar_map_table['rhs'] = 'bar_longer'
+
+    eq(
+      {foolong_bar_map_table, foo_bar_map_table},
+      nvim('get_keymap', 'n')
+    )
+
+    -- Remove a mapping
+    nvim('command', 'unmap foo_longer')
+    eq(
+      {foo_bar_map_table},
+      nvim('get_keymap', 'n')
+    )
+  end)
+
+  it('works for other modes', function()
+    -- Add two mappings, one in insert and one normal
+    -- We'll only check the insert mode one
+    nvim('command', 'nnoremap not_going to_check')
+
+    nvim('command', 'inoremap foo bar')
+    -- The table will be the same except for the mode
+    local insert_table = local_copy(foo_bar_map_table)
+    insert_table['mode'] = 'i'
+
+    eq({insert_table}, nvim('get_keymap', 'i'))
+  end)
+
+  it('works for other modes with modifiers', function()
```

Comment:
This talks about “modifiers” but actually checks only `<nowait>`. Should be “… with <nowait>”.